File: play-ws-standalone/src/main/java/play/libs/ws/DefaultBodyWritables.java
/*
 * Copyright (C) Lightbend Inc. <https://www.lightbend.com>
 */

package play.libs.ws;

import akka.stream.javadsl.FileIO;
import akka.stream.javadsl.Source;
import akka.stream.javadsl.StreamConverters;
import akka.util.ByteString;

import java.io.File;
import java.io.InputStream;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.function.Supplier;

/**
 * This interface contains useful {@link BodyWritable} subclasses and default methods.
 *
 * Most of the time, you'll want to implement this interface and then use the static factory methods:
 *
 * <pre>
 * {@code
 * public class MyClient implements DefaultBodyWritables {
 *     public void doStuff() {
 *         wsClient.url("http://example.com").post(body("hello world")).thenApply(response ->
 *             ...
 *         );
 *     }
 * }
 * }
 * </pre>
 *
 * You can always instantiate the writables directly using other methods as necessary:
 *
 * <pre>
 * {@code
 * public class MyClient {
 *     private BodyWritable<String> someOtherMethod(String string) {
 *       return new InMemoryBodyWritable(ByteString.fromString(string), "text/plain");
 *     }
 *
 *     public void doStuff() {
 *         wsClient.url("http://example.com").post(someOtherMethod("hello world")).thenApply(response ->
 *             ...
 *         );
 *     }
 * }
 * }
 * </pre>
 */
public interface DefaultBodyWritables {

    /**
     * Creates a {@link InMemoryBodyWritable} from a String, setting the content-type to "text/plain".
     *
     * @param s the string to pass in.
     * @return a {@link InMemoryBodyWritable} instance.
     */
    default BodyWritable<ByteString> body(String s) {
        return body(s,"text/plain");
    }

    /**
     * Creates a {@link InMemoryBodyWritable} from a String.
     *
     * @param s the string to pass in.
     * @param contentType to pass in.
     * @return a {@link InMemoryBodyWritable} instance.
     */
    default BodyWritable<ByteString> body(String s, String contentType) {
        return new InMemoryBodyWritable(ByteString.fromString(s), contentType);
    }

    /**
     * Creates a {@link InMemoryBodyWritable} from a byte array.
     *
     * @param array the byte array to pass in.
     * @return a {@link InMemoryBodyWritable} instance.
     */
    default BodyWritable<ByteString> body(byte[] array) {
        return body(array, "application/octet-stream");
    }

    /**
     * Creates a {@link InMemoryBodyWritable} from a byte array.
     *
     * @param array the byte array to pass in.
     * @param contentType to pass in.
     * @return a {@link InMemoryBodyWritable} instance.
     */
    default BodyWritable<ByteString> body(byte[] array, String contentType) {
        return new InMemoryBodyWritable(ByteString.fromArray(array), contentType);
    }

    /**
     * Creates a {@link InMemoryBodyWritable} from a byte array using "application/octet-stream"
     * as the content type.
     *
     * @param buffer the byte array to pass in.
     * @return a {@link InMemoryBodyWritable} instance.
     */
    default BodyWritable<ByteString> body(ByteBuffer buffer) {
        return body(buffer, "application/octet-stream");
    }

    /**
     * Creates a {@link InMemoryBodyWritable} from a byte array.
     *
     * @param buffer the byte array to pass in.
     * @param contentType to pass in.
     * @return a {@link InMemoryBodyWritable} instance.
     */
    default BodyWritable<ByteString> body(ByteBuffer buffer, String contentType) {
        return new InMemoryBodyWritable(ByteString.fromByteBuffer(buffer), contentType);
    }

    /**
     * Creates a {@link SourceBodyWritable}, streaming the body incrementally, setting the content-type to "application/octet-stream".
     *
     * @param source the source of byte string.
     * @return a {@link SourceBodyWritable} instance.
     */
    default BodyWritable<Source<ByteString, ?>> body(Source<ByteString, ?> source) {
        return new SourceBodyWritable(source);
    }

    /**
     * Creates a {@link SourceBodyWritable}, setting the content-type to "application/octet-stream".
     *
     * @param file the file
     * @return a {@link SourceBodyWritable} instance.
     */
    default BodyWritable<Source<ByteString, ?>> body(File file) {
        return new SourceBodyWritable(FileIO.fromFile(file));
    }

    /**
     * Creates {@link InMemoryBodyWritable} from form data, setting the content-type to ""application/x-www-form-urlencoded"
     *
     * @param formData the form data
     * @return the serialized form data in form encoded data, using URLEncoder.encode.
     */
    default BodyWritable<ByteString> body(Map<String, String> formData) {
        try {
            List<String> values = new ArrayList<>(formData.size());
            for (Map.Entry<String, String> item : formData.entrySet()) {
                String key = URLEncoder.encode(item.getKey(), "UTF-8");
                String value = URLEncoder.encode(item.getValue(), "UTF-8");
                values.add(key + '=' + value);
            }
            String s = String.join("&", values);
            ByteString byteString = ByteString.fromString(s);
            return new InMemoryBodyWritable(byteString, "application/x-www-form-urlencoded");
        } catch (UnsupportedEncodingException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Creates a {@link SourceBodyWritable}, setting the content-type to "application/octet-stream".
     *
     * @param is the inputstream
     * @return a {@link SourceBodyWritable} instance.
     */
    default BodyWritable<Source<ByteString, ?>> body(Supplier<InputStream> is) {
        return new SourceBodyWritable(StreamConverters.fromInputStream(is::get));
    }

}

File: play-ws-standalone/src/main/scala/play/api/libs/ws/DefaultBodyWritables.scala
/*
 * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
 */

package play.api.libs.ws

import java.io.File
import java.nio.ByteBuffer
import java.util.function.Supplier

import org.apache.pekko.stream.scaladsl.StreamConverters.fromInputStream
import org.apache.pekko.stream.scaladsl.FileIO
import org.apache.pekko.stream.scaladsl.Source
import org.apache.pekko.util.ByteString

import scala.jdk.FunctionConverters._

/**
 * Default BodyWritable for a request body, for use with
 * requests that take a body such as PUT, POST and PATCH.
 *
 * {{{
 * import scala.concurrent.ExecutionContext
 *
 * import play.api.libs.ws.StandaloneWSClient
 * import play.api.libs.ws.DefaultBodyWritables._
 *
 * class MyClass(ws: StandaloneWSClient) {
 *   def postBody()(implicit ec: ExecutionContext) = {
 *     val getBody: String = "..."
 *     ws.url("...").post(getBody).map { response => ??? }
 *   }
 * }
 * }}}
 */
trait DefaultBodyWritables {

  /**
   * Creates an SourceBody with "application/octet-stream" content type from a file.
   */
  implicit val writableOf_File: BodyWritable[File] = {
    BodyWritable(file => SourceBody(FileIO.fromPath(file.toPath)), "application/octet-stream")
  }

  /**
   * Creates an SourceBody with "application/octet-stream" content type from an inputstream.
   */
  implicit val writableOf_InputStream: BodyWritable[Supplier[java.io.InputStream]] = {
    BodyWritable(supplier => SourceBody(fromInputStream(supplier.asScala)), "application/octet-stream")
  }

  /**
   * Creates an SourceBody with "application/octet-stream" content type from a file.
   */
  implicit val writableOf_Source: BodyWritable[Source[ByteString, ?]] = {
    BodyWritable(source => SourceBody(source), "application/octet-stream")
  }

  /**
   * Creates an InMemoryBody with "text/plain" content type.
   */
  implicit val writeableOf_String: BodyWritable[String] = {
    BodyWritable(str => InMemoryBody(ByteString.fromString(str)), "text/plain")
  }

  /**
   * Creates an InMemoryBody with "text/plain" content type from a StringBuilder
   */
  implicit val writeableOf_StringBuilder: BodyWritable[StringBuilder] = {
    BodyWritable(str => InMemoryBody(ByteString.fromString(str.toString())), "text/plain")
  }

  /**
   * Creates an InMemoryBody with "application/octet-stream" content type from an array of bytes.
   */
  implicit val writeableOf_ByteArray: BodyWritable[Array[Byte]] = {
    BodyWritable(bytes => InMemoryBody(ByteString(bytes)), "application/octet-stream")
  }

  /**
   * Creates an InMemoryBody with "application/octet-stream" content type from a bytebuffer.
   */
  implicit val writeableOf_ByteBuffer: BodyWritable[ByteBuffer] = {
    BodyWritable(buffer => InMemoryBody(ByteString.fromByteBuffer(buffer)), "application/octet-stream")
  }

  /**
   * Creates an InMemoryBody with "application/octet-stream" content type.
   */
  implicit val writeableOf_Bytes: BodyWritable[ByteString] = {
    BodyWritable(byteString => InMemoryBody(byteString), "application/octet-stream")
  }

  /**
   * Creates a BodyWritable with an identity function, with "application/octet-stream" content type.
   */
  implicit val writeableOf_WsBody: BodyWritable[WSBody] = {
    BodyWritable(identity, "application/octet-stream")
  }

  /**
   * Creates an InMemoryBody with "application/x-www-form-urlencoded" content type.
   */
  implicit val writeableOf_urlEncodedForm: BodyWritable[Map[String, Seq[String]]] = {
    import java.net.URLEncoder
    BodyWritable(
      formData =>
        InMemoryBody(
          ByteString.fromString(
            formData.flatMap(item => item._2.map(c => s"${item._1}=${URLEncoder.encode(c, "UTF-8")}")).mkString("&")
          )
        ),
      "application/x-www-form-urlencoded"
    )
  }

  implicit val writeableOf_urlEncodedSimpleForm: BodyWritable[Map[String, String]] = {
    writeableOf_urlEncodedForm.map[Map[String, String]](_.map(kv => kv._1 -> Seq(kv._2)))
  }

}

object DefaultBodyWritables extends DefaultBodyWritables
